### ISSUE TITLE
Schema inference for JSONEachRow returns inconsistent results across versions

### ISSUE THREAD

**Issue by DerekChia** [ClickHouse/ClickHouse#44361]
**Describe the unexpected behaviour**
Reading ndjson file as `JSONEachRow` format using different ClickHouse versions would return different results.

**How to reproduce**
Query:
```
DESCRIBE s3('https://datasets-documentation.s3.eu-west-3.amazonaws.com/github/github-2022.ndjson.gz', 
'JSONEachRow') SETTINGS input_format_max_rows_to_read_for_schema_inference=100;
```
Reference: https://clickhouse.com/docs/en/guides/developer/working-with-json/json-semi-structured/#relying-on-schema-inference

| Version                                       | Output                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          |
|-----------------------------------------------|-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| 22.3.15.33                                    | Received exception from server (version 22.3.15):<br><br>Code: 636. DB::Exception: Received from localhost:9000. DB::Exception: Cannot extract table structure from JSONEachRow format file. Error: Cannot determine table structure by first 100 rows of data, because some columns contain only Nulls. To increase the maximum number of rows to read for structure determination, use setting input_format_max_rows_to_read_for_schema_inference. (CANNOT_EXTRACT_TABLE_STRUCTURE)<br>(query: DESCRIBE s3('https://datasets-documentation.s3.eu-west-3.amazonaws.com/github/github-2022.ndjson.gz', <br>'JSONEachRow') SETTINGS input_format_max_rows_to_read_for_schema_inference=100;)                                                                                                                                                                                                                                                                     |
| 22.4.6.53                                     | ┌─name───────┬─type─────────────┬─default_type─┬─default_expression─┬─comment─┬─codec_expression─┬─ttl_expression─┐<br>│ type       │ Nullable(String) │              │                    │         │                  │                │<br>│ actor      │ Object('json')   │              │                    │         │                  │                │<br>│ repo       │ Object('json')   │              │                    │         │                  │                │<br>│ created_at │ Nullable(String) │              │                    │         │                  │                │<br>│ payload    │ Object('json')   │              │                    │         │                  │                │<br>└────────────┴──────────────────┴──────────────┴────────────────────┴─────────┴──────────────────┴────────────────┘<br><br>5 rows in set. Elapsed: 1.680 sec.                                                         |
| 22.5.4.19 / 22.6.9.11 / 22.7.7.24 / 22.9.7.34 | ┌─name───────┬─type─────────────────────┬─default_type─┬─default_expression─┬─comment─┬─codec_expression─┬─ttl_expression─┐<br>│ type       │ Nullable(String)         │              │                    │         │                  │                │<br>│ actor      │ Object(Nullable('json')) │              │                    │         │                  │                │<br>│ repo       │ Object(Nullable('json')) │              │                    │         │                  │                │<br>│ created_at │ Nullable(String)         │              │                    │         │                  │                │<br>│ payload    │ Object(Nullable('json')) │              │                    │         │                  │                │<br>└────────────┴──────────────────────────┴──────────────┴────────────────────┴─────────┴──────────────────┴────────────────┘<br><br>5 rows in set. Elapsed: 1.690 sec. |
| 22.10.4.23                                    | Received exception from server (version 22.10.4):<br>Code: 361. DB::Exception: Received from localhost:9000. DB::Exception: Seek position is out of bounds. Offset: 0, Max: 0: Cannot extract table structure from JSONEachRow format file. You can specify the structure manually. (SEEK_POSITION_OUT_OF_BOUND)                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                |
| 22.8.11.15 / 22.11.2.30 / 22.12.1.1752        | Received exception from server (version 22.12.1):<br>Code: 652. DB::Exception: Received from localhost:9000. DB::Exception: Cannot determine type for column 'actor' by first 100 rows of data, most likely this column contains only Nulls or empty Arrays/Maps. You can specify the type for this column using setting schema_inference_hints: Cannot extract table structure from JSONEachRow format file. You can specify the structure manually. (ONLY_NULLS_WHILE_READING_SCHEMA)                                                                                                                                                                                                                                                                                                                                                                                                                                                                         |

Also for 22.8.11.15 / 22.11.2.30 / 22.12.1.1752, the `actor` column in the first row does not contain Nulls or empty Arrays/Maps. 
```
DESCRIBE s3('https://datasets-documentation.s3.eu-west-3.amazonaws.com/github/github-2022.ndjson.gz', 
'JSONEachRow') SETTINGS input_format_max_rows_to_read_for_schema_inference=1;
```

```
Received exception from server (version 22.12.1):
Code: 652. DB::Exception: Received from localhost:9000. DB::Exception: Cannot determine type for column 'actor' by first 1 rows of data, most likely this column contains only Nulls or empty Arrays/Maps. You can specify the type for this column using setting schema_inference_hints: Cannot extract table structure from JSONEachRow format file. You can specify the structure manually. (ONLY_NULLS_WHILE_READING_SCHEMA)
(query: DESCRIBE s3('https://datasets-documentation.s3.eu-west-3.amazonaws.com/github/github-2022.ndjson.gz', 
'JSONEachRow') SETTINGS input_format_max_rows_to_read_for_schema_inference=1;)
```

First row:
```
{"type": "PushEvent", "actor": {"avatar_url": "https://avatars.githubusercontent.com/u/93110249?", "display_login": "Lakshmipatil2021", "id": 93110249, "login": "Lakshmipatil2021", "url": "https://api.github.com/users/Lakshmipatil2021"}, "repo": {"id": 429298592, "name": "revacprogramming/pps-test1-Lakshmipatil2021", "url": "https://api.github.com/repos/revacprogramming/pps-test1-Lakshmipatil2021"}, "created_at": "2022-01-04T07:00:00Z", "payload": {"pull_request": {"updated_at": "", "user": {"login": ""}, "number": 0, "title": "", "state": "", "author_association": "", "head": {"ref": "", "sha": ""}, "base": {"ref": "", "sha": ""}}, "action": "", "ref": "refs/heads/main", "ref_type": "", "size": 1, "distinct_size": 1}}
```

First row working as expected in 22.9.7.34.

```
d245228e5206 :) select version()

SELECT version()

Query id: 7e579dfc-ca97-4494-b814-5ade86a7ba24

┌─version()─┐
│ 22.9.7.34 │
└───────────┘

1 row in set. Elapsed: 0.006 sec.

d245228e5206 :) SELECT actor FROM s3('https://datasets-documentation.s3.eu-west-3.amazonaws.com/github/github-2022.ndjson.gz',
                'JSONEachRow') LIMIT 1 SETTINGS input_format_max_rows_to_read_for_schema_inference=100;

SELECT actor
FROM s3('https://datasets-documentation.s3.eu-west-3.amazonaws.com/github/github-2022.ndjson.gz', 'JSONEachRow')
LIMIT 1
SETTINGS input_format_max_rows_to_read_for_schema_inference = 100

Query id: f814f712-e73f-4d93-aa0f-0d97afbd7de8

┌─actor────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
│ {"avatar_url":"https:\/\/avatars.githubusercontent.com\/u\/93110249?","display_login":"Lakshmipatil2021","id":93110249,"login":"Lakshmipatil2021","url":"https:\/\/api.github.com\/users\/Lakshmipatil2021"} │
└──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘

1 row in set. Elapsed: 14.310 sec.
```

**Expected behavior**
ClickHouse should be able to read ndjson file as expected. LTS versions (22.3, 22.8) should also return consistent results.

**Comment by Avogar** [comment]
Schema inference was introduced only 1 year ago and constantly improving Here the problem is that your data contains type `Object('json')`.
`22.3.15.33` - inference for `Object('json')` is not implemented yet.
`22.4.6.53` - inference for `Object('json')` is implemented.
`22.5.4.19` - we made object type `Nullable` as it's done for other types (it's more natural, because by reading a sample of data we cannot say it column contains nulls or not).
`22.10.4.23` - there was a bug in inference of Dates that was fixed in the next release
`22.10.4.23+` - we disabled inference for `Object('json')` if setting `allow_experimental_object_type` is disabled, because this feature is experimantal.

**Comment by Avogar** [comment]
```
avogar-dev :) DESCRIBE s3('https://datasets-documentation.s3.eu-west-3.amazonaws.com/github/github-2022.ndjson.gz',
              'JSONEachRow') settings allow_experimental_object_type=1

DESCRIBE TABLE s3('https://datasets-documentation.s3.eu-west-3.amazonaws.com/github/github-2022.ndjson.gz', 'JSONEachRow')
SETTINGS allow_experimental_object_type = 1

Query id: 1378b884-e962-45d2-8b97-fd2ec61bff48

┌─name───────┬─type─────────────────────┬─default_type─┬─default_expression─┬─comment─┬─codec_expression─┬─ttl_expression─┐
│ type       │ Nullable(String)         │              │                    │         │                  │                │
│ actor      │ Object(Nullable('json')) │              │                    │         │                  │                │
│ repo       │ Object(Nullable('json')) │              │                    │         │                  │                │
│ created_at │ Nullable(String)         │              │                    │         │                  │                │
│ payload    │ Object(Nullable('json')) │              │                    │         │                  │                │
└────────────┴──────────────────────────┴──────────────┴────────────────────┴─────────┴──────────────────┴────────────────┘

5 rows in set. Elapsed: 0.532 sec.

avogar-dev :) select version()

SELECT version()

Query id: 43c78e4c-b864-421d-97dc-996708712eeb

┌─version()───┐
│ 22.12.1.520 │
└─────────────┘

1 row in set. Elapsed: 0.001 sec.
```

**Comment by Avogar** [comment]
So the only inconsistence I see is that we don't respect `allow_experimental_object_type` in versions `22.4-22.9` @DerekChia dou you think we should backport this change
?

**Comment by Avogar** [comment]
Also error message is not good in this case, need to mention setting `allow_experimental_object_type`. I already did it in this PR: https://github.com/ClickHouse/ClickHouse/pull/44019/

**Comment by DerekChia** [comment]
> Also error message is not good in this case, need to mention setting `allow_experimental_object_type`. I already did it in this PR: #44019

Great, that would be a good reminder. 

> So the only inconsistence I see is that we don't respect `allow_experimental_object_type` in versions `22.4-22.9` @DerekChia do you think we should backport this change?

Thanks, didn't occur to me that I need to set `allow_experimental_object_type` since some versions are working. 

I tried again with `allow_experimental_object_type=1` on [fiddle](https://fiddle.clickhouse.com/) and got the error on 22.3, 22.8 and 22.10. I think we should at least backport to 22.3 since it is LTS. 

```
DESCRIBE s3('https://datasets-documentation.s3.eu-west-3.amazonaws.com/github/github-2022.ndjson.gz',
              'JSONEachRow') settings allow_experimental_object_type=1
```

| Version                                                                   | Output                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    |
|---------------------------------------------------------------------------|-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| 22.3.15.33                                                                | Received exception from server (version 22.3.15):<br>Code: 636. DB::Exception: Received from localhost:9000. DB::Exception: Cannot extract table structure from JSONEachRow format file. Error: Cannot determine table structure by first 100 rows of data, because some columns contain only Nulls. To increase the maximum number of rows to read for structure determination, use setting input_format_max_rows_to_read_for_schema_inference. (CANNOT_EXTRACT_TABLE_STRUCTURE)<br>(query: DESCRIBE s3('https://datasets-documentation.s3.eu-west-3.amazonaws.com/github/github-2022.ndjson.gz',<br>              'JSONEachRow') settings allow_experimental_object_type=1<br>)         |
| 22.4.6.53                                                                 | type        Nullable(String)                                        <br>actor        Object(\'json\')                                        <br>repo        Object(\'json\')                                        <br>created_at        Nullable(String)                                        <br>payload        Object(\'json\')                                                                                                                                                                                                                                                                                                                                                    |
| 22.5.4.19 / 22.6.9.11 / 22.7.7.24 / 22.9.7.34 / 22.11.2.30 / 22.12.1.1752 | type        Nullable(String)                                        <br>actor        Object(Nullable(\'json\'))                                        <br>repo        Object(Nullable(\'json\'))                                        <br>created_at        Nullable(String)                                        <br>payload        Object(Nullable(\'json\'))                                                                                                                                                                                                                                                                                                                      |
| 22.8.11.15                                                                | Received exception from server (version 22.8.11):<br>Code: 652. DB::Exception: Received from localhost:9000. DB::Exception: Cannot determine type for column payload by first 25000 rows of data, most likely this column contains only Nulls or empty Arrays/Maps. You can specify the type for this column using setting schema_inference_hints: Cannot extract table structure from JSONEachRow format file. You can specify the structure manually. (ONLY_NULLS_WHILE_READING_SCHEMA)<br>(query: DESCRIBE s3('https://datasets-documentation.s3.eu-west-3.amazonaws.com/github/github-2022.ndjson.gz',<br>              'JSONEachRow') settings allow_experimental_object_type=1<br>) |
| 22.10.4.23                                                                | Received exception from server (version 22.10.4):<br>Code: 361. DB::Exception: Received from localhost:9000. DB::Exception: Seek position is out of bounds. Offset: 0, Max: 0: Cannot extract table structure from JSONEachRow format file. You can specify the structure manually. (SEEK_POSITION_OUT_OF_BOUND)<br>(query: DESCRIBE s3('https://datasets-documentation.s3.eu-west-3.amazonaws.com/github/github-2022.ndjson.gz',<br>              'JSONEachRow') settings allow_experimental_object_type=1<br>)                                                                                                                                                                          |

**Comment by Avogar** [comment]
> I think we should at least backport to 22.3 since it is LTS.

You ment 22.8? Because 22.3 just doesn't containt feature with inference for json object.
I will fix errors on 22.8 and 22.10 and backport respecting setting `allow_experimental_object_type` to make it consistent.
